### DISCUSSION TITLE
[MAUI, Android] Using INotificationManager from foreground service (IGpsDelegate.OnReading)

### DISCUSSION THREAD

**Discussion by ChilloDE** [shinyorg/shiny#1275] [Q&A] [answered]
Hello, 

I thought before I create a minimal PoC and write a full bugreport, I'd ask here first if this may be an expected behavior on android, that I just missed in the docs and at least can't seem to find any infromation about.

Installed ShinyPackages: 
Shiny.Hosting.Maui
Shiny.Locations
Shiny.Notifications
Shiny.Push (Currently not set up in MauiProgramm.cs, just the reference)
All are Version 3.0.0-beta-0179

Initialisation:

```
var builder = MauiApp.CreateBuilder();
builder
    .UseMauiApp<App>()
    .UseShiny()
// [...]
builder.Services.AddGps<BackgroundLocationDelegate>();
builder.Services.AddNotifications();

return builder.Build();`
```

I did not register a INotificationDelegate for now. As far as I understand the samples this is only for callbacks from clicking on notifications?

I've got the IGpsDelegate implementation and Android foreground service running (with a notification to inform the user about the activity) fetching the devices position and uploding it to a webservice in short intervalls, this works fine.
Since I am already communicating with my webservice and only need to send notifications while the location tracking is active anyway I though instead of going through firebase as a 3rd party provider, I could just use the response of my webservice to inform the client about updates. And show them via local notifications. I've made sure to create a channel and add it to the notification. But the INotificationManager.Send-method seems to never return and notifications are not displayed, when I use Send inside the implementation of IGpsDelegate.OnReading.

Broken down I'm not doing much more than:
```
public async Task OnReading(GpsReading reading)
{
    var data = _restClient.Position(reading);
    if(data != null)
    {
        await _notificationManager.Send(new Notification()
        {
            Title = data.Title,
            Text = data.Text,
            Thread = Strings.WarningsThread,
            Channel = Strings.NotificationChannel,
        }); // <--- This call does not return.
    }
}
```

Long story short is it even possible to send local notifications fom the gps foreground service delegate on android?

**Comment by aritchie** [comment]
You have to run INotificationManager.RequestAccess in the foreground to be able to send.

**Reply by aritchie** [reply]
You also need to create your custom channel

**Reply by ChilloDE** [reply]
Thanks for the reply.
When I looked back on my code I realized the error was a simple one on my part, I missed to set a reference to _notificationManager in the constructor. (I should stop trying to debug my code late at night)

I got it to work now, thanks again.

The code that is calling IGpsDelegate.OnReading seems to be dismissing any uncaught (NullReference)Exceptions though.
I am handling the following uncaught excepition handlers:
AppDomain.CurrentDomain.UnhandledException
TaskScheduler.UnobservedTaskException
AndroidEnvironment.UnhandledExceptionRaiser

I just went back and checked it with breakpoints in all Event-handlers, none of them was raised nor did the app crash.

**Reply by aritchie** [reply, marked as answer]
Logging is built into Shiny.  You need to hook up to the Microsoft Logging Extensions.  You can a file logger and Shiny also provides a sqlite logger for tracking things in the background

**Comment by magtimmermans** [comment]
Hi, I am trying to do the same with the foreground service from android and log the location every x minutes. However, I get this not working. The service starts fine. Sometimes the OnGPSreading will be triggered but mostly not. Would you maybe willing to check my Example?
[https://github.com/magtimmermans/TestLocations](https://github.com/magtimmermans/TestLocations)

Thanks in advance!

**Reply by aritchie** [reply]
@magtimmermans This is against our terms & conditions.  Please don't solicit users for support like this on our forum.  GPS is working fine.  I'm not sure why you don't like my answer.  I don't do anything to block GPS pings.  They either work or they don't.  I'm using them right now.

**Reply by magtimmermans** [reply]
Okay, clear. Unfortunately it won't work here.
